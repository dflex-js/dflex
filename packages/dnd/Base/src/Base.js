import { DRAGGED_ELM } from "@dflex/draggable/constants.json";
import AbstractDraggable from "@dflex/draggable/src/AbstractDraggable";

import store from "../../Store";

import { ACTIVE_PARENT } from "../../constants.json";

/**
 * Base element.
 *
 * Creates draggedElm and activeParent and initializes thresholds.
 *
 *
 * @class Base
 */
class Base extends AbstractDraggable {
  /**
   * Creates an instance of Base.
   *
   * @param {string} elementId
   * @memberof Base
   */
  constructor(elementId, clickCoordinates) {
    const {
      element,
      parent,
      branches: { siblings, parents },
    } = store.getElmTreeById(elementId);

    super(element, clickCoordinates);

    const { order } = element;

    /**
     * Initialize temp index that refers to element new position after
     * transformation happened.
     */
    this.draggedTempIndex = order.self;

    this.parentsList = parents;
    this.siblingsList = siblings;

    /**
     * Thresholds store, contains max value for each parent and for dragged. Depending on
     * ids as keys.
     */
    this.thresholds = {
      parents: {},
      dragged: {},
    };

    this.checkThresholds(this[DRAGGED_ELM]);

    /**
     * Init max direction for position
     */
    this.setThreshold(this[DRAGGED_ELM], false);

    this.setIsSingleton();

    this.setIsOrphan(parent);
  }

  /**
   * Check if dragged has no siblings and then set the related flag.
   *
   * @memberof Base
   */
  setIsSingleton() {
    /**
     * Dragged has no siblings.
     */
    this.isSingleton = !Array.isArray(this.siblingsList);
  }

  /**
   * Check if dragged has no parent and then set the related operations
   * accordingly.
   *
   * @memberof Base
   */
  setIsOrphan(parent) {
    /**
     * Dragged has no parent.
     */
    this.isOrphan = true;

    /**
     * Not all elements have parents.
     */
    if (parent) {
      /**
       * Indicator to parents that have changed. This facilitates looping in
       * affected parents only.
       */
      this.setOfTransformedIds = new Set([]);
      this.assignActiveParent(parent);
      this.setThreshold(parent, true);
      this.isOrphan = false;
    }
  }

  /**
   * Sets thresholds for dragged element position depending on its
   * position inside parent which is related to droppable left and top.
   *
   * maxDirection = current position + droppable-allowed spaces
   *
   * @memberof Base
   */
  setThreshold(droppable, isParent) {
    const { parents, dragged } = this.thresholds;

    const {
      thresholdOffset: { vertical, horizontal },
    } = this[DRAGGED_ELM];

    const {
      currentLeft,
      currentTop,
      offset: { width, height },
      id,
    } = droppable;

    const $ = isParent ? (parents[id] = {}) : dragged;

    /**
     * Overview:
     *
     * this.maxDirection: contains max value allowed for dragged to move. If dragged passes
     * these values, that means it's out.
     *
     * This threshold is related to droppable offset. If droppable changes, then
     * max/threshold changes.
     *
     * parentThreshold: to detect if dragged is moved inside or outside
     * parent. But what if we have multiple lists?
     *
     * Updating positionMax when dragged is located in new position is easy but
     * in lists we need to know all listMax(s) before entering to parent. Why?
     * because in this way we know where dragged is to detect the activeParent.
     *
     * The solution is to create parentThreshold. An object stores positionMax
     * depending on currentIndex as keys.
     */

    /**
     * Calculate max-vertical for up and down:
     */

    /**
     * When going up, currentTop decreases (-vertical).
     */
    $.maxTop = currentTop - vertical;

    /**
     * When going down, currentTop increases (+vertical) with droppable
     * taking into considerations (+ vertical).
     */
    $.maxBottom = currentTop + vertical;

    /**
     * When going left, currentLeft decreases (-horizontal).
     */
    $.maxLeft = currentLeft - horizontal;

    /**
     * When going right, currentLeft increases (+horizontal) with droppable
     * taking into considerations (+ horizontal).
     */
    $.maxRight = currentLeft + horizontal;
  }

  /**
   * Check thresholds availability and assign it to thresholds instance in
   * dragging process.
   *
   * @memberof Base
   */
  checkThresholds() {
    /**
     * Not initiating thresholdOffset for all coreInstances. Only ones which
     * dragged will be initiated.
     */
    if (!this[DRAGGED_ELM].thresholdOffset) {
      const {
        offset: { width, height },
      } = this[DRAGGED_ELM];

      /**
       * Calculates thresholdOffset only for dragged element.
       *
       * Two-thirds of the dragged element's space for vertical and horizontal. If
       * two-thirds of the dragged is out, then trigger isOut whether it is out
       * position or out parent.
       */
      this[DRAGGED_ELM].thresholdOffset = {
        vertical: Math.ceil((2 / 3) * height),
        horizontal: Math.ceil((2 / 3) * width),
      };
    }
  }

  /**
   * Add parent id to setOfTransformedIds. The goal here, is to avoid any
   * unnecessary process the parent goes through it when it's flagged as
   * transformed.
   *
   * @memberof Base
   */
  addParentAsTransformed() {
    const { id } = this[ACTIVE_PARENT];

    /**
     * Avoid adding same parents more than once using sets.
     */
    this.setOfTransformedIds.add(id);
  }

  /**
   * Assigns new ACTIVE_PARENT: parent who contains dragged
   *
   * @param {CoreInstance} coreInstance
   * @memberof Base
   */
  assignActiveParent(coreInstance) {
    /**
     * Assign instance ACTIVE_PARENT which represents droppable. Then
     * assign owner parent so we have from/to.
     */
    this[ACTIVE_PARENT] = coreInstance;

    /**
     * Add flag for undo method so we can check which  parent is being
     * transformed and which is not.
     */
    this[ACTIVE_PARENT].isTransformed = true;
    this.isDraggedOutActiveParent = false;

    /**
     * In initiating we get siblingsList. But, when dragged is in another
     * ACTIVE_PARENT we have different siblingsList now. siblingsList is
     * children of new parent. Anyway, check if there no list, get us one.
     */
    if (!this.siblingsList) {
      const {
        keys: { chK },
      } = this[ACTIVE_PARENT];

      this.siblingsList = store.getElmById(chK);
    }
  }
}

export default Base;
