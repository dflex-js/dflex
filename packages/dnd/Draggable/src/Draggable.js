import { DRAGGED_ELM } from "@dflex/draggable/constants.json";
import Base from "./Base";

class Draggable extends Base {
  constructor(elementId, clickCoordinates) {
    super(elementId, clickCoordinates);

    const { x, y } = clickCoordinates;

    this.innerOffsetX = x - this[DRAGGED_ELM].currentLeft;
    this.innerOffsetY = y - this[DRAGGED_ELM].currentTop;

    this.tempOffset = {
      currentLeft: 0,
      currentTop: 0,
    };

    /**
     * previous X and Y are used to calculate mouse directions.
     */
    this.prevX = x;
    this.prevY = y;

    this.numberOfElementsTransformed = 0;

    /**
     * Elements effected by dragged direction.
     */
    this.elemDirection = 1;

    this.isMovingDownPrev = false;
    this.isMovingDown = false;
  }

  /**
   * Dragged current-offset is essential to determine dragged position in
   * layout and parent.
   *
   * Is it moved form its translate? Is it out the parent or in
   * another parent? The answer is related to currentOffset.
   *
   * Note: these are the current offset related only to the dragging. When the
   * operation is done, different calculation will be set.
   *
   * @memberof Draggable
   */
  dragAt(x, y) {
    this.translate(x, y);

    /**
     * Every time we got new translate, offset should be updated
     */
    this.tempOffset.currentLeft = x - this.innerOffsetX;
    this.tempOffset.currentTop = y - this.innerOffsetY;
  }

  /**
   * Checks if dragged it out of its position or parent.
   *
   * @param {this|this.parentThreshold[currentIndex]} $
   * @returns {boolean} isOut
   * @memberof Draggable
   */
  isDraggedOut(id) {
    const { parents, dragged } = this.thresholds;

    const $ = id ? parents[id] : dragged;

    const isOut =
      this.tempOffset.currentLeft < $.maxLeft ||
      this.tempOffset.currentLeft > $.maxRight ||
      this.tempOffset.currentTop < $.maxTop ||
      this.tempOffset.currentTop > $.maxBottom;

    return isOut;
  }

  /**
   * Checks if dragged is last element in parent list.
   *
   * @returns {boolean}
   * @memberof Draggable
   */
  isDraggedLastElm() {
    return this.isSingleton || this.tempIndex === this.siblingsList.length - 1;
  }

  /**
   * Checks if dragged is first element in parent list.
   *
   * @returns {boolean}
   * @memberof Draggable
   */
  isDraggedFirstElm() {
    return this.isSingleton || this.tempIndex === 0;
  }

  /**
   * Checks if dragged is moving down and updates element direction sign (+/-).
   *
   * @returns {boolean}
   * @memberof Draggable
   */
  updateDraggedDirectionFlags(y) {
    this.isMovingDownPrev = this.isMovingDown;
    this.isMovingDown = y > this.prevY;
    this.prevY = y;

    if (
      this.numberOfElementsTransformed !== 0 &&
      this.isMovingDownPrev !== this.isMovingDown
    ) {
      /**
       * In this case, we have a sudden change in mouse movement. So, reverse
       * numberOfElementsTransformed value, to be compatible with elemDirection.
       */
      this.numberOfElementsTransformed *= -1;
    }

    /**
     * If dragged is going top, element will decrease. So:
     * Down: -1, up: 1.
     */
    this.elemDirection = this.isMovingDown ? -1 : 1;
  }

  endDragging() {
    this.setDragged(false);

    this[DRAGGED_ELM].setCurrentOffset();
  }
}

export default Draggable;
