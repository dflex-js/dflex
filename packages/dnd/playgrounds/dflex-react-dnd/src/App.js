/* eslint-disable no-plusplus */
import React from "react";
import "./App.css";

import Core from "./Core";
import Container from "./Container";

const firstContainer = [{ label: "container1 |> elm-1", id: "1" }];

const secondContainer = [
  { label: "container2 |> elm-1", id: "2" },
  { label: "container2 |> elm-2", id: "3" },
  { label: "container2 |> elm-3", id: "4" },
  { label: "container2 |> elm-4", id: "5" },
  { label: "container2 |> elm-5", id: "6" },
  { label: "container2 |> elm-6", id: "7" },
  { label: "container2 |> elm-7", id: "8" },
];

const thirdContainer = [
  { label: "container3 |> elm-1", id: "9" },
  { label: "container3 |> elm-2", id: "10" },
  { label: "container3 |> elm-3", id: "11" },
  { label: "container3 |> elm-4", id: "12" },
];

function App() {
  return (
    <Container className="Container">
      <Core id="p0-a" component="ul" depth={1}>
        {firstContainer.map(({ label, id }) => (
          <Core depth={0} id={`id-${id}`} key={`k${id}`} component="li">
            {label}
          </Core>
        ))}
      </Core>
      <Core id="p0-1b" component="ul" depth={1}>
        {secondContainer.map(({ label, id }) => (
          <Core depth={0} id={`id-${id}`} key={`k-${id}`} component="li">
            {label}
          </Core>
        ))}
      </Core>
      <Core id="p0-1c" component="ul" depth={1}>
        {thirdContainer.map(({ label, id }) => (
          <Core depth={0} id={`id-${id}`} key={`k-${id}`} component="li">
            {label}
          </Core>
        ))}
      </Core>
    </Container>
  );
}

export default App;
