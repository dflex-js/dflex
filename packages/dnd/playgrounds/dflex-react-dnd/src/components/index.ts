/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
export { default as Depth1 } from "./depth-1";
export { default as ExtendedList } from "./extended";

export {
  RestrictedSelf,
  RestrictedContainerAll,
  RestrictedContainerDiff,
} from "./restrictions";

export { TodoList, TodoListCase1 } from "./todo";

export { default as UnRegisteredLists } from "./mixedRegisteredElements";

export { ComponentBasedEvent, ContainerBasedEvent } from "./essential";

export { ScrollMulti, ScrollEssential } from "./scroll";
