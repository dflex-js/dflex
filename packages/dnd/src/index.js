import DnD from "./DnD";
import store from "./DnDStoreImp";

export { DnD, store };
