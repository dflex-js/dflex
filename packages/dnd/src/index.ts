export { default as store } from "./DnDStore";
export { default as DnD } from "./DnD";
