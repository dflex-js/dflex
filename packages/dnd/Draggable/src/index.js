import Draggable from "./Draggable";

export default Draggable;
