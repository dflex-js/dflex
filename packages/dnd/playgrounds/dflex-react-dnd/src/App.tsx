/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { store } from "@dflex/dnd";

import {
  Restricted,
  TodoList,
  TodoListCase1,
  ExtendedList,
  Depth1,
  ComponentBasedEvent,
  ContainerBasedEvent,
} from "./components";

function App() {
  React.useEffect(() => {
    return () => {
      store.destroy();
    };
  }, []);

  return (
    <Router>
      <Switch>
        <Route path="/extended">
          <ExtendedList />
        </Route>
        <Route path="/restricted">
          <Restricted />
        </Route>
        <Route path="/todo">
          <TodoList />
        </Route>
        <Route path="/todo-case-1">
          <TodoListCase1 />
        </Route>
        <Route path="/depth-1">
          <Depth1 />
        </Route>
        <Route path="/component-based-event">
          <ComponentBasedEvent />
        </Route>
        <Route path="/">
          <ContainerBasedEvent />
        </Route>
      </Switch>
    </Router>
  );
}

export default App;
