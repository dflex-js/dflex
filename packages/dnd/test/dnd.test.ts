import { store, DnD } from "../src/index";

import { elm1, elm2, elm3, elm4 } from "./utils";

describe("Checks Store Instance", () => {
  beforeAll(() => {
    store.register(elm1);
    store.register(elm2);
    store.register(elm3);
    store.register(elm4);
  });

  it("Dragging first element to the end of the list", () => {
    const dnd = new DnD(elm1.id, { x: 551, y: 122 });
    for (let i = 122; i <= 286; i += 1) {
      dnd.dragAt(551, i);
    }

    dnd.endDragging();
  });

  it("Checks offset of elements", () => {
    const offsetElm1 = store.getELmOffsetById(elm1.id);
    const offsetElm2 = store.getELmOffsetById(elm2.id);
    const offsetElm3 = store.getELmOffsetById(elm3.id);
    const offsetElm4 = store.getELmOffsetById(elm4.id);

    expect(offsetElm1).toMatchSnapshot();
    expect(offsetElm2).toMatchSnapshot();
    expect(offsetElm3).toMatchSnapshot();
    expect(offsetElm4).toMatchSnapshot();
  });

  it("Checks transition of elements", () => {
    const transElm1 = store.getELmTranslateById(elm1.id);
    const transElm2 = store.getELmTranslateById(elm2.id);
    const transElm3 = store.getELmTranslateById(elm3.id);
    const transElm4 = store.getELmTranslateById(elm4.id);

    expect(transElm1).toMatchSnapshot();
    expect(transElm2).toMatchSnapshot();
    expect(transElm3).toMatchSnapshot();
    expect(transElm4).toMatchSnapshot();
  });

  it("Checks the order of siblings", () => {
    const siblings = store.getElmSiblingsById(elm1.id);

    expect(siblings).toStrictEqual(["id-2", "id-3", "id-4", "id-1"]);
  });

  it("Dragging first element to the top of the list", () => {
    const dnd = new DnD(elm1.id, { x: 551, y: 286 });

    for (let i = 286; i >= 122; i -= 1) {
      dnd.dragAt(551, i);
    }

    dnd.endDragging();
  });

  it("Restore offset of elements", () => {
    const offsetElm1 = store.getELmOffsetById(elm1.id);
    const offsetElm2 = store.getELmOffsetById(elm2.id);
    const offsetElm3 = store.getELmOffsetById(elm3.id);
    const offsetElm4 = store.getELmOffsetById(elm4.id);

    expect(offsetElm1).toMatchSnapshot();
    expect(offsetElm2).toMatchSnapshot();
    expect(offsetElm3).toMatchSnapshot();
    expect(offsetElm4).toMatchSnapshot();
  });

  it("Restore transition of elements", () => {
    const transElm1 = store.getELmTranslateById(elm1.id);
    const transElm2 = store.getELmTranslateById(elm2.id);
    const transElm3 = store.getELmTranslateById(elm3.id);
    const transElm4 = store.getELmTranslateById(elm4.id);

    expect(transElm1).toMatchSnapshot();
    expect(transElm2).toMatchSnapshot();
    expect(transElm3).toMatchSnapshot();
    expect(transElm4).toMatchSnapshot();
  });

  it("Makes sure the order of siblings is back in correct order", () => {
    const siblings = store.getElmSiblingsById(elm1.id);

    expect(siblings).toStrictEqual(["id-1", "id-2", "id-3", "id-4"]);
  });
});
