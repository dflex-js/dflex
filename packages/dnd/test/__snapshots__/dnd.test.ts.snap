// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checks Store Instance Checks offset of elements 1`] = `
Object {
  "height": 50,
  "left": 450,
  "top": 288,
  "width": 170,
}
`;

exports[`Checks Store Instance Checks offset of elements 2`] = `
Object {
  "height": 50,
  "left": 450,
  "top": 114,
  "width": 170,
}
`;

exports[`Checks Store Instance Checks offset of elements 3`] = `
Object {
  "height": 50,
  "left": 450,
  "top": 172,
  "width": 170,
}
`;

exports[`Checks Store Instance Checks offset of elements 4`] = `
Object {
  "height": 50,
  "left": 450,
  "top": 230,
  "width": 170,
}
`;

exports[`Checks Store Instance Checks transition of elements 1`] = `
Object {
  "translateX": 0,
  "translateY": 174,
}
`;

exports[`Checks Store Instance Checks transition of elements 2`] = `
Object {
  "translateX": 0,
  "translateY": -58,
}
`;

exports[`Checks Store Instance Checks transition of elements 3`] = `
Object {
  "translateX": 0,
  "translateY": -58,
}
`;

exports[`Checks Store Instance Checks transition of elements 4`] = `
Object {
  "translateX": 0,
  "translateY": -58,
}
`;

exports[`Checks Store Instance Restore offset of elements 1`] = `
Object {
  "height": 50,
  "left": 450,
  "top": 288,
  "width": 170,
}
`;

exports[`Checks Store Instance Restore offset of elements 2`] = `
Object {
  "height": 50,
  "left": 450,
  "top": 114,
  "width": 170,
}
`;

exports[`Checks Store Instance Restore offset of elements 3`] = `
Object {
  "height": 50,
  "left": 450,
  "top": 172,
  "width": 170,
}
`;

exports[`Checks Store Instance Restore offset of elements 4`] = `
Object {
  "height": 50,
  "left": 450,
  "top": 230,
  "width": 170,
}
`;

exports[`Checks Store Instance Restore transition of elements 1`] = `
Object {
  "translateX": 0,
  "translateY": 0,
}
`;

exports[`Checks Store Instance Restore transition of elements 2`] = `
Object {
  "translateX": 0,
  "translateY": 0,
}
`;

exports[`Checks Store Instance Restore transition of elements 3`] = `
Object {
  "translateX": 0,
  "translateY": 0,
}
`;

exports[`Checks Store Instance Restore transition of elements 4`] = `
Object {
  "translateX": 0,
  "translateY": 0,
}
`;
